numpy/ma/mrecords.py

```diff
@@ -668,7 +668,8 @@ def openfile(fname):
 
 
 def fromtextfile(fname, delimiter=None, commentchar='#', missingchar='',
-                 varnames=None, vartypes=None):
+                 varnames=None, vartypes=None,
+                 *, delimitor=None):  # mispelled for backwards compatibility
```

Comment:
I don't think this quite works.  Suppose I used `fromtextfile("a.txt", ",")`.  Then I will get a warning.  It is a real pain to do this completely correctly (I think it requires moving to *args, and **kwargs and then a lot of manual parsing.  You might simple add to the warning string when delimiter is not None that end users can use the keyword argument delimitor=value to silence the warning, since you don't know if they really misspelled it or not.

Reply:
The only problem I see is if the user passes `delimitor=None`.  The solution to which is using `np._NoValue` rather than None?  I assume we have tests that would fail if `fromtextfile("a.txt", ",")` gave a spurious warning?

Reply:
I don't get a warning with the following code:
```python
def fromtextfile(fname, delimiter=None, *, delimitor=None):
    if delimitor is not None:
        print('WARNING: "delimitor" should be None!')

fromtextfile("a.txt", delimiter=",")
fromtextfile("a.txt", ",")
```

Reply:
I do get the warning with:
```python
fromtextfile("a.txt", delimitor=",")
```

Reply:
And I get an exception with:
```python
fromtextfile("a.txt", delimiter=",", ",")
fromtextfile("a.txt", ",", ",")
```

Reply:
> The solution to which is using np._NoValue rather than None?

I think the only way to 100% know that you didn't get both is to use **kwargs.  Using np._NoValue is the same as using any other impossible value as a sentinal (e.g., -1).  It still does not stop a user from writing 

```
fromtextfile("a.txt", delimiter=None, delimitor=np._NoValue)
```

while if **kwargs was uses, one could look into kwargs and see is delimiter was set.

Reply:
> I think the only way to 100% know that you didn't get both is to use

Right, but users are not supposed to use `np._NoValue`, it is a private singleton used only for this type of thing (or almost only).  Frankly, if a user passes `np._NoValue` it is their problem, not ours.

Reply:
This emits the warning, then raises the TypeError exception as suggested by @seberg, as expected:
```python
fromtextfile("a.txt", delimiter=",", delimitor=",")
```

Reply:
I don't get it. I think it's easy to see if you have both arguments: test `delimiter` and `delimitor`.

Reply:
> This emits the warning, then raises the TypeError exception as suggested by @seberg, as expected:

Shouldn't the error have precedence over the warning?

Reply:
> I don't get it. I think it's easy to see if you have both arguments: test `delimiter` and `delimitor`.

You can't normally tell the difference between

```
fromtextfile("a.txt")
fromtextfile("a.txt", None)
fromtextfile("a.txt", delimiter=None)
fromtextfile("a.txt", delimitor=None)
fromtextfile("a.txt", delimiter=None, delimitor=None)
fromtextfile("a.txt", None, delimitor=None)
```

Unless you do terrible things using `*args` and `**kwargs`.  There might be introspection tools that can also tell but it is quite a heavy lift.

Reply:
```python
fromtextfile("a.txt", delimiter=None, delimitor=np._NoValue)
```
Well, I haven't tried this, but do we really want to spend time shielding against the following use of a function that doesn't even seem to be used out there (see usage statistics in published code).
```python
fromtextfile("a.txt", delimiter=None, delimitor=np._NoValue)

Reply:
Actually the following does emit a warning:
```python
fromtextfile("a.txt", delimiter=None, delimitor=np._NoValue)
```

Reply:
The only case where we do **not** get a warning is:
```python
fromtextfile("a.txt", delimiter=",", delimitor=None)
```
Perhaps that's the use case you have in mind?

Reply:
I thought it would be nice to have both the warning, then the exception raised. But I get it, that's too verbose and should be avoided. I'll revert to what @seberg had suggested.

Reply:
So the use case you have in mind is indeed:
```python
fromtextfile("a.txt", delimitor=None)
```
It should emit a warning but doesn't.

Reply:
I can try to change it to **kwargs, but I am getting tired of this. I wasn't expecting such long an contradictory debates.

Reply:
Now the code is equivalent to:
```python
import numpy as np

def fromtextfile(fname, delimiter=None, *, delimitor=np._NoValue):
    if delimitor is not np._NoValue:
        print('WARNING: "delimitor" should not be used any more!')
```
This will now emit a warning:
```python
fromtextfile("a.txt", delimitor=None)
```
but indeed the following will **not** emit a warning:
```python
fromtextfile("a.txt", delimitor=np._NoValue)
```

Given the usage statistics of `fromtextfile()`, I wouldn't worry about it.

On the other hand, it would be nice to have a robust (enough) mechanism to handle deprecation of keyword arguments. I just don't think I'm the person to write such a mechanism and get it to be as foolproof as possible. At the very least that should happen in a separate PR or even in a module of its own. I haven't found a relevant module, but I have found a few articles:
* [How to implement deprecation in python with argument alias](https://stackoverflow.com/questions/49802412/how-to-implement-deprecation-in-python-with-argument-alias#49802489)
* [Deprecate a function parameter](https://stackoverflow.com/questions/54949421/deprecate-a-function-parameter#54949947)

Reply:
I could add decorators such as `@deprecate_keyword_argument` and `@rename_keyword_argument`, for example under in `lib/utils.py`alongside the `_Deprecate` decorator class, or wherever you feel is more appropriate. But again, I'd rather do that in a different PR.

Reply:
Please don't, we have used this trick for similar things already, it is nice and simple, and I am not aware of a (reasonable) failure mode.